addons/stock/models/stock_picking.py

```diff
@@ -1031,6 +1075,34 @@ def _check_immediate(self):
                 immediate_pickings |= picking
         return immediate_pickings
 
+    def _check_zqc(self):
+        """ Return the locations that will be emptied out when `self` is processed. """
+        empty_locations = self.env['stock.location']
+
+        # Group move lines per locations
+        ml_per_loc = defaultdict(lambda: self.env['stock.move.line'])
+        for ml in self.move_line_ids:
+            if ml.location_id.usage in ('internal', 'transit'):
+                ml_per_loc[ml.location_id] |= ml
+
+        # Check if locations would be emptied by those movelines
+        for loc, move_line_ids in ml_per_loc.items():
+            quants = self.env['stock.quant'].search([
+                ('location_id', '=', loc.id),
+            ])
+
+            remaining_qties = defaultdict(lambda: 0)
+            for quant in quants:
+                remaining_qties[quant.product_id] += quant.quantity
+            for ml in move_line_ids:
+                remaining_qties[ml.product_id] -= ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
```

Comment:
Can use a `search_read` instead of a `search`:
```suggestion
            quants = self.env['stock.quant'].search_read([
                ('location_id', '=', loc.id),
            ], ['product_id', 'quantity'])

            remaining_qties = defaultdict(lambda: 0)
            for quant in quants:
                remaining_qties[quant['product_id'][0]] += quant['quantity']
            for ml in move_line_ids:
                remaining_qties[ml.product_id.id] -= ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
```